Kingman prototype

JereKoskela opened the pull request:
Two new functions:
- sim_setup_arg uses 1 child, 1 parent jumps to move all lineages to a single pixel. Spatial parameters and a virtual event location are overwritten so that lineages can only intersect with a single pixel, and are unable to migrate out.
- sim_simulate_arg simulates the arg phase using the two types of transitions we discussed: 2 children, one parent for a merger and 1 child, two parents for a recombination. The different recombination mechanism in the ARG has led to some changes to the sim_generate_parents function, and spatial locations are still set so that every lineage intersects with one single pixel only.

I've tried some trial simulations with small numbers of lineages and both few and many loci, and everything looks good. Valgrind also reports no memory leaks. But I haven't done extensive testing, which would probably be easier with a Python wrapper.

Commit:
Added prototype for panmictic simulation phase

Commit:
Added prototype for panmictic simulation phase

jeromekelleher commented on lib/sim.c:
```diff
@@ -907,12 +923,39 @@ sim_generate_parents(sim_t *self)
+    /* In case of a recombination event in the Kingman phase
+     * we require exactly one recombination point.
+     */
+    if(arg_flag == 1) {
```
Minor style nit: there should be a space between the if and the bracket.

jeromekelleher commented on lib/sim.c:
```diff
@@ -892,14 +904,18 @@ sim_update_parental_ancestry(sim_t *self, unsigned int parent,
 {
     int ret = 0;
-    unsigned int j, parent, previous_locus;
+    unsigned int i = 0, j, parent, previous_locus, break_point = 0;
```
Do these variables need to be initialised here? It's usually a bit clearer if the variables are initialised next to where they're used. For example, we might initiatialise total_rate below just above the while loop so it's easy to see the initial value of the variable without going to the top of the function.

jeromekelleher commented on lib/sim.c:
```diff
@@ -907,12 +923,39 @@ sim_generate_parents(sim_t *self)
+        locus_mapping = (int_map_value_t *) node->item;
+        site_2 = (double)(locus_mapping->key);
+        node = node->next;
+        while(node != NULL) {
```
Style nit agina --- `while (node != NULL)` . Otherwise, it looks like `while` is a function.

jeromekelleher commented on lib/sim.c:
```diff
@@ -907,12 +923,39 @@ sim_generate_parents(sim_t *self)
+        while(node != NULL) {
+            site_1 = site_2;
+            locus_mapping = (int_map_value_t *) node->item;
+            site_2 = (double)(locus_mapping->key);
```
Do we need the brackets around `locus_mapping->key`?

jeromekelleher commented on lib/sim.c:
```diff
@@ -907,12 +923,39 @@ sim_generate_parents(sim_t *self)
+        for(j = 1; j < gap_no; j++) {
+            recomb_rates[j] = recomb_rates[j - 1] + recomb_rates[j] / total_rate;
+        }
+        double coin = gsl_rng_uniform(self->rng);
```
In the rest of the code, variable declarations are at the top of each function. It would be nice to be consistent, but it's not a big deal.

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1247,102 @@ sim_simulate(sim_t *self, uint64_t max_events)
+    individual_t *ind;
+    /* An arbitrary radius and parental location for using existing spatial 
+     * functions to complete events */
+    double z[] = {self->pixel_size / 2.0, self->pixel_size / 2.0}, r = self->pixel_size / 4.0;
```
I'd prefer to keep initialisations and declarations seperate. For this sort of thing I think it's too easy to miss an initialisation or set it to the wrong value if they're mixed on the same line. I would just split these onto two lines.

Also, are these inits supported in c99? I don't really want to use c11 features yet, as it makes portability a pain.

JereKoskela replied:
I don't know about c99, but to me this looks identical to declarations and initialisations of the analogous z[] variables at the top of sim_simulate and sim_non_reproduction_event. Could you be more specific about what it should look like instead? Just having z[] and r declared on separate lines?

jeromekelleher replied:
Yeah, sorry, you're right. I just checked (compiling with `-std=c99`) and these sort of declarations are fine. I would prefer to see them on separate lines though, as it makes the initilisation clearer:

```
double z[] = {self->pixel_size / 2.0, self->pixel_size / 2.0};
double r = self->pixel_size / 4.0;
```

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1247,102 @@ sim_simulate(sim_t *self, uint64_t max_events)
+            if (ret != 0) {
+                goto out;
+            }
+        }
```
Join this line and the next:

```
} else {
```

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1247,102 @@ sim_simulate(sim_t *self, uint64_t max_events)
+        self->time += gsl_ran_exponential(self->rng, 1.0 / total_rate);
+        if(gsl_rng_uniform(self->rng) < coal_rate / total_rate) {
+            // Coalescence event
+            self->num_parents = 1;
```
Why are we setting the number of parents here? I think this is used as a constant simulation parameter value in the rest of the simulation, so I don't really like setting it here...

JereKoskela replied:
The Kingman simulation always needs exactly 1 parent for a merger and 2 parents for a recombination, regardless of how many parents there are in the spatial simulation. That's why I set the number of parents explicitly before calling sim_complete_event() in either case. The only alternative I can think of is to add more flag variables to further branch the behaviour of sim_complete_event() depending on the state of the simulation, but this way feels cleaner.

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1247,102 @@ sim_simulate(sim_t *self, uint64_t max_events)
+    unsigned int pixel_occ, i, j;
+    self->num_parents = 1;
+    self->num_children = 1;
+    double z[] = {self->pixel_size / 2.0, self->pixel_size / 2.0}, r = self->pixel_size / 4.0;
```
For older compilers, we should keep all declarations at the start of the function. Also, I'm nervous about setting `self->num_parents` like this...

JereKoskela replied:
I'm using single parent, single child jumps to move all remaining lineages into the first pixel, so as to avoid having to update the avl trees manually. Overwriting the number of parents like this seemed like the easiest way to reuse existing functions. I don't think it's problematic since the original parent number is not needed by the Kingman simulation, but it's possible I'm missing something. What makes you nervous about it?

jeromekelleher replied:
Yes, that's true... I guess I'm just nervous as these are treated as parameters in the spatial simulation and modifying parameters makes me nervous. You're right though, this is probably the simplest way to achieve this and reusing the existing infrastructure as much as possible.

jeromekelleher commented:
This is superb, thanks @JereKoskela! I think this is a great framework, and we'll be able to build the remaining functionality pretty easily on top of this. There are some minor changes I'd like to see:
1. Some binary files have been included in the git commit. These should be removed.
2. I think we're using some pretty recent C features which might cause trouble for portability. C99 is the version I'd like to target.
3. I've made some superficial style comments also, just to try to keep the style consistent through the codebase.

Commit:
Minor style updates

Commit:
Minor style tweaks

JereKoskela commented:
I think I caught all the style discrepancies and removed the binary files. I also split up the z[] and r inits like we discussed. Self->num_parents still gets modified as before. Let me know if you spot anything else.

jeromekelleher commented on lib/.directory:
```diff
@@ -0,0 +1,4 @@
+[Dolphin]
+Timestamp=2015,6,24,17,30,18
+Version=3
+ViewMode=1
```
This file should be removed as well (KDE eh?).

jeromekelleher replied:
Also need to remove 'main' from the commit.

JereKoskela replied:
Both done

Commit:
Removed unnecessary files

jeromekelleher commented on lib/main.c:
```diff
@@ -270,6 +281,10 @@ run_sim(const char *config_file)
         }
         not_done = ret != 0; 
     }
+    if(self->simulate_kingman == 1) {
```
space needed here - if (..)

jeromekelleher commented on lib/sim.c:
```diff
@@ -173,11 +173,9 @@ index_to_pixel_coord(int index, unsigned int N, unsigned int *v)
 {
-    double ret = gsl_sf_choose(n, k);
-    ret *= gsl_pow_int(u, k) * gsl_pow_int(1.0 - u, n - k);
+    double ret = exp(gsl_sf_lnchoose(n, k) + (double)(k) * log(u) + (double)(n - k) * log(1.0 - u));
```
Why are we changing this? Are we hitting overflow issues with the Kingman phase or something?

JereKoskela replied:
Yes. Tests with thousands of loci tended to cause overflow, and this solved the problem.

jeromekelleher replied:
Cool, nice fix. I think we can get rid of the (double) casts here, as they'll be upcast automatically. In general, it's good to keep lines to ~80 chars.

jeromekelleher commented on lib/sim.c:
```diff
@@ -562,6 +569,7 @@ sim_add_pixel_to_occupancy(sim_t *self, unsigned int h, unsigned int pixel)
             goto out;
         }
         if (avl_insert_node(tree, h_node) == NULL) {
+            printf("Insert node error");
```
Are these printfs necessary? We can keep them for now if they're useful for debugging, but they should be cleared out once we're happy everything's working OK.

JereKoskela replied:
This is odd. I can see the line in my Github repository, but not in the sim.c file in my local branch. I'll try and investigate how it got here.

jeromekelleher commented on lib/sim.c:
```diff
@@ -1239,6 +1401,8 @@ sim_print_parameters(sim_t *self)
         printf("# \tr = %f\n", self->event_classes[j].r);
         printf("# \trate = %f\n", self->event_classes[j].rate);
     }
+    printf("Kingman Ne = %f\n", self->Ne);
```
Should start these lines with a # for consistency.

jeromekelleher commented on lib/main.c:
```diff
@@ -270,6 +281,10 @@ run_sim(const char *config_file)
         not_done = ret != 0; 
     }
+    if(self->simulate_kingman == 1) {
+        ret = sim_setup_arg(self);
```
Need to check these return values for errors: `if (ret != 0)` etc.

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1249,100 @@ sim_simulate(sim_t *self, uint64_t max_events)
+            int_ptr = (uintptr_t) id;
+            ind = (individual_t *) int_ptr;
+            self->child_buffer[0] = ind;
+            ret = sim_complete_event(self, z, r, 0, 1);
```
This seems a bit roundabout --- why not call `sim_remove_individual()` followed by `sim_add_individual` with a new location of (0, 0)? This would mean we don't need the `transition_flag` any more, and would simplify the `sim_complete_event` function.

JereKoskela replied:
Yes, looking at what sim_complete_event actually does, that would be more direct. I'll change it.

JereKoskela replied:
Tracing ancestral relationships through the transition correctly makes this a little more complicated than just removing and adding individuals. But I can probably make it work with a call to `sim_update_parental_ancestry()` before anything gets removed or added.

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1249,100 @@ sim_simulate(sim_t *self, uint64_t max_events)
+    self->num_parents = 1;
+    self->num_children = 1;
+    for (j = 0; j < self->N * self->N; j++) {
+        pixel_occ = avl_count(&self->P[j]);
```
This doesn't work I think --- we're looking at the same individual `pixel_occ` times. Try this (from `sim_get_population`):

``` c
for (node = self->P[pixel].head; node != NULL; node = node->next) 
```

JereKoskela replied:
I tried your suggestion first, and it throws a segmentation fault because self->P[pixel] is altered at each iteration of the for loop. My implementation looks at a new individual each time because the previous head is no longer occupying the pixel. I'll add a comment to that effect, because I agree it isn't completely transparent.

Incidentally, what you suggest is exactly what happens at pixel 0. But I should have skipped that entirely because any lineages that only overlap pixel 0 are already in the right place, and any lineages overlapping pixel 0 and at least one other pixel will be caught in the other pixel.

jeromekelleher replied:
Ahh, _very_ sneaky! A comment would be good here all right.

I don't think we should worry too much about individuals being caught in pixels any more. If we keep the definitions of the two main Kingman functions separate from the spatial methods, then it should all work out nicely with all individuals having a location of (0,0). Once everyone is in pixel 0, that's all we care about (we might even reconfigure the pixels so there is only one; may or may not be worthwhile).

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1249,100 @@ sim_simulate(sim_t *self, uint64_t max_events)
+    if (self->dimension == 1) {
+        z[1] = 0.0;
+    }
+    self->num_parents = 1;
```
OK, so if we're calling `sim_remove_individual` manually here, we don't need to update these parameters any more, right?

JereKoskela replied:
Correct.

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1249,100 @@ sim_simulate(sim_t *self, uint64_t max_events)
+    if (self->dimension == 1) {
+        z[1] = 0.0;
+    }
+    while (self->ancestral_material > self->num_loci) {
```
defining `n = (set_map_value_t *)self->Q.head->item)->key` would improve readability here.

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1249,100 @@ sim_simulate(sim_t *self, uint64_t max_events)
+        coal_rate = 2.0 * self->Ne * gsl_sf_choose(((set_map_value_t *)self->Q.head->item)->key, 2);
+        recomb_rate = 0.0;
+        S_size = 0;
+        for(node = self->P[0].head; node != NULL; node = node->next) {
```
Space between for and ( here. It would be best to just search for `for(` and `if(` throughout.

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1249,100 @@ sim_simulate(sim_t *self, uint64_t max_events)
+            id = *((uint64_t *) node->item);
+            int_ptr = (uintptr_t) id;
+            ind = (individual_t *) int_ptr;
+            recomb_rate += (double)(((set_map_value_t *)ind->ancestry.tail->item)->key 
```
We need some variables here --- what is `(((set_map_value_t *)ind->ancestry.tail->item)->key`? I guess this is `first_locus` and `last_locus`?

jeromekelleher commented on lib/sim.c:
```diff
@@ -1181,7 +1249,100 @@ sim_simulate(sim_t *self, uint64_t max_events)
+            // Coalescence event
+            self->num_parents = 1;
+            self->num_children = 2;
+            gsl_ran_choose(self->rng, self->child_buffer, self->num_children, S, S_size, sizeof(individual_t *));
```
Line is too long here.

jeromekelleher commented on lib/sim.c:
```diff
@@ -907,6 +923,34 @@ sim_generate_parents(sim_t *self)
+        locus_mapping = (int_map_value_t *) node->item;
+        site_2 = (double) locus_mapping->key;
+        node = node->next;
+				total_rate = 0.0;
```
Indentation.

jeromekelleher commented on lib/sim.c:
```diff
@@ -907,6 +923,34 @@ sim_generate_parents(sim_t *self)
+            recomb_rates[j] = recomb_rates[j - 1] + recomb_rates[j] / total_rate;
+        }
+        coin = gsl_rng_uniform(self->rng);
+				break_point = 0;
```
Indentation.

jeromekelleher commented on lib/sim.c:
```diff
@@ -892,14 +904,18 @@ sim_update_parental_ancestry(sim_t *self, unsigned int parent,
     int_map_value_t *locus_mapping;
+    double total_rate, site_1, site_2, coin;
+    gap_no = avl_count(&C[0]->ancestry) - 1;
+    double recomb_rates[gap_no];
```
This is dodgy --- allocating anything variable length off the stack can blow up very nastily. A good pattern to follow here is

``` c
double recomb_rates = NULL;
recomb_rates = xmalloc(gap_no * sizeof(double));

// Then, later on
out:
    if (recomb_rates != NULL) {
        free(recomb_rates);
    }
```

This involves a malloc for every event, but we can make a buffer for that later if we need to.

jeromekelleher commented on lib/sim.c:
```diff
@@ -907,6 +923,34 @@ sim_generate_parents(sim_t *self)
+            total_rate += recomb_rates[i];
+            i++;
+            node = node->next;
+        } 
```
So, what we're doing here is iterating over all the mapped loci in this individual, and calculating the probability that a recombination occurs between two pairs, right? I think it would be good to phrase this explicitly as probablilities, and reuse the `probability_list_select` function above to choose the location we're interested in.

jeromekelleher commented on lib/sim.c:
```diff
@@ -921,11 +965,29 @@ sim_generate_parents(sim_t *self)
                 goto out;
             }
             previous_locus = locus_mapping->key;
+						i = 0;
```
indentation.

jeromekelleher commented on lib/sim.c:
```diff
@@ -921,11 +965,29 @@ sim_generate_parents(sim_t *self)
+                if(arg_flag == 0) {
+                    parent = sim_select_parent(self, parent, locus_mapping->key
                         - previous_locus);
+                }
```
```
} else {
    /*Track the recombination point in the Kingman phase*/
```

jeromekelleher commented on lib/sim.c:
```diff
@@ -892,14 +904,18 @@ sim_update_parental_ancestry(sim_t *self, unsigned int parent,
+
 static int 
-sim_generate_parents(sim_t *self)
+sim_generate_parents(sim_t *self, int arg_flag)
```
I think it would be better to write a new function, `sim_generate_arg_parents`, and leave `sim_geneate_parents` as it is (we might even rename it `sim_generate_spatial_parents`). We're making both parts much more complicated by combining them I think. 

In fact, I think it would be easier if we made two new functions that just did the required Kingman stuff: 
`sim_arg_coancestry_event()` and `sim_arg_recombination_event`. We can still reuse the basic bits like `sim_remove_individual`, `sim_add_individual` and `sim_update_parental_ancestry` which should make it relatively straightforward.

jeromekelleher commented:
OK, I've had a more detailed look through this now, and it's clearer what's going on. You've done a great job integrating in with the existing code, but I wonder if perhaps you've gone a little too far in this. I think it would simplify things a lot if we made top level functions to perform the two types of event we want in the Kingman phase. It would be a lot easier to follow what's happening then, and we wouldn't need to trace through a bunch of functions using flags to determine what code path to take. 

Does this make sense?

Commit:
Separated out spatial and ARG simulations

jeromekelleher commented on lib/sim.c:
```diff
@@ -943,6 +955,172 @@ sim_generate_parents(sim_t *self)
+     */
+    int i, break_point = -2;
+    double *recomb_rates = NULL;
+    recomb_rates = xmalloc(gap_no * sizeof(double));
```
Need to check if this is null here:

``` c
if (recomb_rates == NULL) {
    ret = ERR_NO_MEMORY;
    goto out;
}
```

jeromekelleher commented on lib/sim.c:
```diff
@@ -584,6 +591,7 @@ sim_remove_pixel_from_occupancy(sim_t *self, unsigned int h, unsigned int pixel)
         smv_search.key = h;
         q_node = avl_search(&self->Q, &smv_search);
         if (q_node == NULL) {
+            printf("q_node error");
```
We should remove these printfs at some point.
